### PR TITLE
[Mips] Fix unable to handle inline assembly ends with compat-branch o…

### LINKED ISSUE #61045: Unable to handle inline assembly ends with compat-branch on MIPS
Currently LLVM unable to handle the situation that a inline Assembly ends with branch instruction and then compiler generate another CTI immediately.

It would leave compiler generated CTI in forbidden slot and crash at runtime.

Test case:

```ll
define i32 @foo0() nounwind {
entry:
  %0 = tail call i32 asm "1: addiu $0, $$0, 1; beqzc $0, 1b;", "=r"() nounwind
  ret i32 %0
}
```

Generate object file with:
```shell
llc -mtriple=mips64el-linux-gnuabi64 -mcpu=mips64r6 -filetype=obj ./broken-forbidden-slot.ll
```

Then dissemble with:

```console
$ mips64el-linux-gnuabi64-objdump -d ./broken-forbidden-slot.o
./broken-forbidden-slot.o:     file format elf64-tradlittlemips


Disassembly of section .text:

0000000000000000 <foo0>:
   0:   24020001        li      v0,1
   4:   d85ffffe        beqzc   v0,0 <foo0>
   8:   d81f0000        jrc     ra
```

JRC lies in forbidden slot of BEQZC.

### PR DESCRIPTION
**Pull request by yingopq**
…n MIPS

Modify:
Add a global variable 'CurForbiddenSlotAttr' to save current instruction's forbidden slot and whether set reorder. This is the judgment condition for whether to add nop. We would add a couple of '.set noreorder' and '.set reorder' to wrap the current instruction and the next instruction.
Then we can get previous instruction`s forbidden slot attribute and whether set reorder by 'CurForbiddenSlotAttr'.
If previous instruction has forbidden slot and .set reorder is active and current instruction is CTI. Then emit a NOP after it.

Fix https://github.com/llvm/llvm-project/issues/61045.

Because https://reviews.llvm.org/D158589 was 'Needs Review' state, not ending, so we commit pull request again.

### CODE REVIEW

**Comment by yingopq** [comment]
@topperc Could you help review this patch? Thanks.

**Comment by yingopq** [comment]
@topperc Hello, could you help review this patch at your convenience? Thanks.

**Comment by yingopq** [comment]
@MaskRay 
Hi, this issue is https://reviews.llvm.org/D158589, which you have reviewed before.
Could you help review this patch at your convenience? Thanks.

**Comment by jrtc27** [comment]
What assembly does GCC emit, and how does GNU as deal with that assembly? That is, how do those two tools interact in the GNU world when dealing with forbidden slots?

**Comment by yingopq** [comment]
> What assembly does GCC emit, and how does GNU as deal with that assembly? That is, how do those two tools interact in the GNU world when dealing with forbidden slots?

All results were same: add nop after beqzc.
```
$ cat a.s
	.set reorder
bbb:
	beqzc $t1, bbb
	b bbb
$ sudo mipsel-linux-gnu-as -mips32r6 a.s
$ mipsel-linux-gnu-objdump -d a.out

a.out:     file format elf32-tradlittlemips


Disassembly of section .text:

00000000 <bbb>:
   0:	d93fffff 	beqzc	t1,0 <bbb>
   4:	00000000 	nop
   8:	1000ffff 	b	8 <bbb+0x8>
   c:	00000000 	nop
```

and
```
$ sudo mipsel-linux-gnu-gcc -mips32r6 -c a.s -o hello.o
$ mipsel-linux-gnu-objdump -d hello.o 

hello.o:     file format elf32-tradlittlemips


Disassembly of section .text:

00000000 <bbb>:
   0:	d93fffff 	beqzc	t1,0 <bbb>
   4:	00000000 	nop
   8:	1000ffff 	b	8 <bbb+0x8>
   c:	00000000 	nop
```

**Comment by yingopq** [comment]
> What assembly does GCC emit, and how does GNU as deal with that assembly? That is, how do those two tools interact in the GNU world when dealing with forbidden slots?

All results were same: add nop after beqzc.
```
$ cat a.s
	.set reorder
bbb:
	beqzc $t1, bbb
	b bbb
$ sudo mipsel-linux-gnu-as -mips32r6 a.s
$ mipsel-linux-gnu-objdump -d a.out

a.out:     file format elf32-tradlittlemips


Disassembly of section .text:

00000000 <bbb>:
   0:	d93fffff 	beqzc	t1,0 <bbb>
   4:	00000000 	nop
   8:	1000ffff 	b	8 <bbb+0x8>
   c:	00000000 	nop
```

and
```
$ sudo mipsel-linux-gnu-gcc -mips32r6 -c a.s -o hello.o
$ mipsel-linux-gnu-objdump -d hello.o 

hello.o:     file format elf32-tradlittlemips


Disassembly of section .text:

00000000 <bbb>:
   0:	d93fffff 	beqzc	t1,0 <bbb>
   4:	00000000 	nop
   8:	1000ffff 	b	8 <bbb+0x8>
   c:	00000000 	nop
```

**Comment by yingopq** [comment]
@wzssyqa 
Hi, this issue is https://reviews.llvm.org/D158589, which you have reviewed before.
Could you help review this patch? Thanks.

File: llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp [hunk 1]
```diff
@@ -2307,7 +2319,42 @@ bool MipsAsmParser::processInstruction(MCInst &Inst, SMLoc IDLoc,
 
   bool FillDelaySlot =
       MCID.hasDelaySlot() && AssemblerOptions.back()->isReorder();
-  if (FillDelaySlot)
+
+  // Get previous instruction`s forbidden slot attribute and
+  // whether set reorder.
+  // This 'CurForbiddenSlotAttr' is a global variable.
```
Comment (wzssyqa) [inline llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp:2325]: It is not `global` variable, instead it is a class level variable.

File: llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp [hunk 2]
```diff
@@ -2331,6 +2389,14 @@ bool MipsAsmParser::processInstruction(MCInst &Inst, SMLoc IDLoc,
...
   // If this instruction has a delay slot and .set reorder is active,
   // emit a NOP after it.
+  // The format of the third case is:
+  // .set noreorder
+  // bnezc
+  // nop
+  // .set noreorder
+  // j
+  // nop
+  // .set reorder
```
Comment (wzssyqa) [inline llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp:2399]: I think that we should add a test case for the ".set reorder/noreorder" logic.

File: llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp (hunk 1 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp:2325]: OK, I would modify these comments.

**Pull request merged**